src/py21cmfast/drivers/coeval.py

```diff
@@ -814,7 +814,7 @@ def _redshift_loop_generator(
                 if (
                     inputs.matter_options.USE_HALO_FIELD
                     and write.halobox
-                    and iz + 1 < len(all_redshifts)
+                    and iz + 1 < len(inputs.node_redshifts)
```

Comment:
Eh, sorry but I think it's a little more complicated than this. I think we need another counter variable that tracks which node redshift we're up to. `iz` is the the number of redshifts we've completed including the out redshifts, but we have no tracker of the node redshift count. That's what should be used here.